### DISCUSSION TITLE
Tracking with transverse and longitudinal HOMs

### DISCUSSION THREAD

**Discussion by TeresiaOlsson** [atcollab/at#621] [Q&A]
Hi,

I want to simulate transverse and longitudinal HOMs and benchmark the results against what I get with Elegant, but I'm not sure which of the new elements in pyAT that are best to use for this. Ideally I would like to use the phasor addition method and be able to extract the state of the phasor each turn.

Is it just BeamLoadingElement that can use the phasor method and which saves the state of the phasor? I looked at the ResonatorElement but it seems to only use a wake and I didn't manage to find any saved data for which voltage the beam sees each turn.

Best regards,

Teresia

**Comment by swhite2401** [comment]
Hi Teresia,

Could you give us some more details on what you are trying to model? There might be different options whether you are looking at your main accelerating cavity or harmonic cavities passive / active.

For information, @lcarver has done some benchmarking with elegant and we have found some differences in the calculation of the beam voltage. In pyAT we output the average Vbeam while elegant seems to output the value just before the bunch passage, we are waiting for M. Borland's confirmation on that point, @lcarver may give you some more information on this.

**Reply by TeresiaOlsson** [reply]
I want to simulate a general narrow resonator. It can be a higher order mode in any type of cavity (active/passive) or a narrow resonator from other type of impedance element (like a BPM). In Elegant it can be done with phasor addition using the elements FRFMODE (longitudinal) or TRFMODE (transverse). So for this I don't need all the complexity related to beam loading compensation and RF feedback, but I would like to be able to output the amplitude and phase of the voltage induced by the impedance to compare with what Elegant gives and what I expect analytically. But perhaps there is no element to do that yet in pyAT, so I need to do it with a wake?

For FRFMODE and TRFMODE I think Elegant doesn't give so great output, but Elegant also have the element RFMODE (which is used for a single longitudinal mode and can be run together with RF feedback). For that one you can get two different voltages, "seen by bunch" or "after bunch". I'm guessing that "seen by bunch" should be some kind of average voltage that the bunch sees, but I'm not sure how it's actually calculated. Personally, I always instead use the "after bunch" voltage because it describes the cavity state and if you know the coordinates of the beam you can always construct what the beam should see from that. @lcarver Is RFMODE the element you have benchmarked against in Elegant? If so, I can also test it and see if I get a difference. I have benchmarked Elegant against mbtrack and analytic formulas for harmonic cavities before and I usually had good agreement.

**Reply by swhite2401** [reply]
The wake method will not output the beam voltage so I am not sure you can really compare the results. Have you considered looking at particles coordinates before and after the resonator element?

For the beamloading element you can set the feedback gain to zero and have zero accelerating voltage to be in "passive" mode. You may add as many such elements as HOMs and you will get the complex vbeam output for each of them.
I have never tried it but it could be working, this would be for the longitudinal plane only.

If you think such phasor resonator sum would be useful in pyAT it can be added. We had discussed this possibility but I thought there was not so much interest in it, apparently I was wrong!

**Reply by TeresiaOlsson** [reply]
Okay, I guessed it was the case that it didn't exist yet. I can use the beam loading element to benchmark the longitudinal to start with. The reason I want to be able to see the voltage and not only look at the particle coordinates is because I have concerns about how Elegant models the radiation. Elegant seems to only have a 4D orbit and not 6D and I have experienced major issues when having dispersion at the observation point (which I unfortunately have). So to figure out if it's the beam dynamics model in Elegant which is doing something funny or if there is a difference in the impedance it would help with the voltage.

I think it would be useful with an element like that. In Elegant I have used it to simulate growth rates of longitudinal coupled-bunch instabilities caused by HOMs in the cavities or transverse coupled-bunch instabilites caused by IDs and collimators. Then it helps with a fast simulation method and to not have to specify how many turns to save for the wake.

**Reply by swhite2401** [reply]
Ok thanks for the feedback, will be put in the TODO list

**Reply by lcarver** [reply]
The wake method will not output the beam voltage. But will output enough information to compute the beam voltage yourself (in the turn history). However for high Q resonators like you mention, the phasor method is much more appropriate and we don't have it implemented with the LongResonatorElement or TransResonatorElement. 

I agree with Simon that it is a worthwhile addition. In the meantime, I wonder if you can add a BeamLoadingElement at high frequency with VoltGain and PhaseGain set to 0. Only thing to be careful of is maintaining the right frequency as the psi may detune the frequency secretly. There is a development branch called add_passive_cavity (i think). Maybe this can be used. But it has not been benchmarked yet which is why it is not merged with master. I am away at the moment so I cannot check the code. Either way it doesn't help for transverse case.
